### PR TITLE
Create method for UsersController and Test

### PR DESCRIPTION
**Pull request by SakhPrace**
#360
- [X] Add create method for UsersController
- [X] Add tests for create method for UsersController

### CODE REVIEW

**Comment by SakhPrace** [comment]
@malinink check this PR pls.

**Comment by SakhPrace** [comment]
@malinink I'm not sure about exception we should throw when we try to create user with existing email. And should we test this in special test like " createEmailFailedTest"?

File: src/main/java/easytests/api/v1/controllers/UsersController.java [hunk 1]
```diff
@@ -29,9 +35,24 @@
...
-     * create
-     */
+    @PostMapping("")
+    @ResponseStatus(HttpStatus.CREATED)
+    public Identity create(@RequestBody User user) throws BadRequestException {
+        if (user.getId() != null) {
+            throw new IdentifiedModelException();
+        }
+        if (this.usersService.findByEmail(user.getEmail()) != null) {
+            throw new BadRequestException();
```
Comment (malinink) [inline src/main/java/easytests/api/v1/controllers/UsersController.java]: I think that we should add some additional comments in that case, that e-mail already present

File: src/main/java/easytests/api/v1/controllers/UsersController.java [hunk 2]
```diff
@@ -29,9 +35,24 @@
...
+        }
+
+        final UserModelInterface userModel = this.usersMapper.map(user, UserModel.class);
+
+        userModel.setPassword("");
```
Comment (malinink) [inline src/main/java/easytests/api/v1/controllers/UsersController.java]: lets generate random password (6 letters, find how to do that)

File: src/test/java/easytests/api/v1/controllers/UsersControllerTest.java [hunk 3]
```diff
@@ -1,23 +1,138 @@
...
-     * testCreateWithSubjectsFailed
-     */
+
+     @Test
+     public void testCreateSuccess() throws Exception {
+         doAnswer(invocation -> {
+             final UserModel userModel = (UserModel) invocation.getArguments()[0];
+             userModel.setId(5);
+             return null;
+         }).when(this.usersService).save(any(UserModelInterface.class));
```
Comment (malinink) [inline src/test/java/easytests/api/v1/controllers/UsersControllerTest.java]: mmm, I miss that in Object create test, but we need to verify that object mapped correct.
So we need to get userModel from support class, and verify that passed user equals userModel.

**Review by malinink — CHANGES_REQUESTED** [review]

**Comment by SakhPrace** [comment]
@malinink Pls check this PR.

File: src/test/java/easytests/api/v1/controllers/UsersControllerTest.java (hunk 3 above)
Comment (SakhPrace) [reply src/test/java/easytests/api/v1/controllers/UsersControllerTest.java]: @malinink Comment my solution. I don't sure.

**Comment by SakhPrace** [comment]
@malinink Check pls

File: src/test/java/easytests/api/v1/controllers/UsersControllerTest.java [hunk 4]
```diff
@@ -110,15 +117,81 @@ public void testListForbidden() throws Exception {
...
-     */
-    /**
-     * testCreateWithSubjectsFailed
-     */
+    @Test
+     public void testCreateSuccess() throws Exception {
+         doAnswer(invocation -> {
+             final UserModel userModel = (UserModel) invocation.getArguments()[0];
+             userModel.setId(5);
+             final UserEntity userTrueEntity = this.usersSupport.getEntityAdditionalMock(0);
```
Comment (malinink) [inline src/test/java/easytests/api/v1/controllers/UsersControllerTest.java]: What are the proposals?
Only you need is to set Id here.

File: src/test/java/easytests/api/v1/controllers/UsersControllerTest.java [hunk 5]
```diff
@@ -110,15 +117,81 @@ public void testListForbidden() throws Exception {
...
+             final UserModel userTrueModel = new UserModel();
+             userTrueModel.map(userTrueEntity);
+             userTrueModel.setId(5);
+             userTrueModel.setPassword(userModel.getPassword());
+             userTrueModel.setSubjects(null);
+             this.usersSupport.assertEquals(userModel, userTrueModel);
+             return null;
+         }).when(this.usersService).save(any(UserModelInterface.class));
+         mvc.perform(post("/v1/users")
```
Comment (malinink) [inline src/test/java/easytests/api/v1/controllers/UsersControllerTest.java]: And catch User model that have been passed into users service, than verify that passed user model consists of correct fields. (mapped from Post data.)
Do no check password, cause of it is generated field. Other - should be checked.

File: src/test/java/easytests/api/v1/controllers/UsersControllerTest.java [hunk 6]
```diff
@@ -40,13 +42,18 @@
     private static String email = "email";
     private static String isAdmin = "isAdmin";
     private static String state = "state";
+    private static String subjects = "subjects";
```
Comment (malinink) [inline src/test/java/easytests/api/v1/controllers/UsersControllerTest.java]: unused subjects

**Review by malinink — CHANGES_REQUESTED** [review]

**Comment by SakhPrace** [comment]
@malinink Ready for review

**Comment by SakhPrace** [comment]
@malinink Ready For Review

File: src/main/java/easytests/api/v1/controllers/UsersController.java [hunk 7]
```diff
@@ -54,6 +61,26 @@ private Boolean isAdmin() {
...
      * create
      */
+
+    @PostMapping("")
+    @ResponseStatus(HttpStatus.CREATED)
+    public Identity create(@RequestBody User user) throws BadRequestException {
+        if (user.getId() != null) {
+            throw new IdentifiedModelException();
+        }
+        if (this.usersService.findByEmail(user.getEmail()) != null) {
```
Comment (malinink) [inline src/main/java/easytests/api/v1/controllers/UsersController.java]: is admin check

File: src/test/java/easytests/api/v1/controllers/UsersControllerTest.java [hunk 8]
```diff
@@ -110,15 +119,82 @@ public void testListForbidden() throws Exception {
...
+                 .andExpect(content().json(
+                         new JsonSupport()
+                                 .with(id, 5)
+                                 .build()
+                 ))
+                 .andReturn();
+         verify(this.usersService, times(1)).save(userCaptor.capture());
+        Assert.assertEquals(userCaptor.getValue().getFirstName(), "FirstName");
+        Assert.assertEquals(userCaptor.getValue().getLastName(), "LastName");
+        Assert.assertEquals(userCaptor.getValue().getSurname(), "SurName");
```
Comment (malinink) [inline src/test/java/easytests/api/v1/controllers/UsersControllerTest.java]: email skipped

File: src/test/java/easytests/api/v1/controllers/UsersControllerTest.java [hunk 9]
```diff
@@ -110,15 +119,82 @@ public void testListForbidden() throws Exception {
...
+                        .with(state, 0)
+                        .build()
+                ))
+                .andExpect(status().isBadRequest())
+                .andExpect(content().string(""))
+                .andReturn();
+    }
+
+    @Test
+    public void testCreateWithSubjectsFailed() throws Exception {
```
Comment (malinink) [inline src/test/java/easytests/api/v1/controllers/UsersControllerTest.java]: remove

**Review by malinink — CHANGES_REQUESTED** [review]

**Comment by SakhPrace** [comment]
@malinink rdy for review. What about userService cant find user by email?

**Comment by SakhPrace** [comment]
@malinink Finally ready for review

File: src/test/java/easytests/api/v1/controllers/UsersControllerTest.java [hunk 10]
```diff
@@ -113,15 +118,138 @@ public void testListForbidden() throws Exception {
...
+                        .with(state, 0)
+                        .build()
+                ))
+                .andExpect(status().isBadRequest())
+                .andExpect(content().string(""))
+                .andReturn();
+    }
+
+    @Test
+    public void testCreateWithIsAdminFailed() throws Exception {
```
Comment (malinink) [inline src/test/java/easytests/api/v1/controllers/UsersControllerTest.java]: `testCreateForbidden`

File: src/test/java/easytests/api/v1/controllers/UsersControllerTest.java [hunk 11]
```diff
@@ -113,15 +118,138 @@ public void testListForbidden() throws Exception {
...
+                        .with(state, 0)
+                        .build()
+                ))
+                .andExpect(status().isBadRequest())
+                .andExpect(content().string(""))
+                .andReturn();
+    }
+
+    @Test
+    public void testCreateWithIsAdminIsUserFailed() throws Exception {
```
Comment (malinink) [inline src/test/java/easytests/api/v1/controllers/UsersControllerTest.java]: `testCreateWithUserIdFailed` ?

**Review by malinink — CHANGES_REQUESTED** [review]

**Comment by SakhPrace** [comment]
@malinink Its done

**Comment by SakhPrace** [comment]
jyjy

**Comment by SakhPrace** [comment]
@malinink Ready for review

**Review by malinink — APPROVED** [review]

**Pull request merged**